File: pdu/python/sensor_msgs/pdu_cdr_conv_Imu.py
from .pdu_pytype_Imu import Imu
from ..pdu_cdr_runtime import CdrReader, CdrWriter

from ..std_msgs.pdu_cdr_conv_Header import *
from ..geometry_msgs.pdu_cdr_conv_Quaternion import *
from ..builtin_interfaces.pdu_cdr_conv_Time import *
from ..geometry_msgs.pdu_cdr_conv_Vector3 import *



def py_to_cdr_body_Imu(writer: CdrWriter, src: Imu):
    py_to_cdr_body_Header(writer, src.header)
    py_to_cdr_body_Quaternion(writer, src.orientation)
    values = src.orientation_covariance
    for i in range(9):
        writer.write_float64(values[i] if i < len(values) else 0.0)
    py_to_cdr_body_Vector3(writer, src.angular_velocity)
    values = src.angular_velocity_covariance
    for i in range(9):
        writer.write_float64(values[i] if i < len(values) else 0.0)
    py_to_cdr_body_Vector3(writer, src.linear_acceleration)
    values = src.linear_acceleration_covariance
    for i in range(9):
        writer.write_float64(values[i] if i < len(values) else 0.0)


def cdr_body_to_py_Imu(reader: CdrReader, dst: Imu):
    cdr_body_to_py_Header(reader, dst.header)
    cdr_body_to_py_Quaternion(reader, dst.orientation)
    dst.orientation_covariance = [reader.read_float64() for _ in range(9)]
    cdr_body_to_py_Vector3(reader, dst.angular_velocity)
    dst.angular_velocity_covariance = [reader.read_float64() for _ in range(9)]
    cdr_body_to_py_Vector3(reader, dst.linear_acceleration)
    dst.linear_acceleration_covariance = [reader.read_float64() for _ in range(9)]
    return dst


def py_to_cdr_Imu(src: Imu) -> bytes:
    writer = CdrWriter()
    writer.write_encapsulation()
    py_to_cdr_body_Imu(writer, src)
    return writer.bytes()


def cdr_to_py_Imu(cdr_payload) -> Imu:
    reader = CdrReader(cdr_payload)
    reader.read_encapsulation()
    dst = Imu()
    return cdr_body_to_py_Imu(reader, dst)

File: pdu/javascript/sensor_msgs/pdu_cdr_conv_Imu.js
import { Imu } from './pdu_jstype_Imu.js';
import { PduCdrWriter, PduCdrReader } from '../pdu_cdr_runtime.js';
import { Header } from '../std_msgs/pdu_jstype_Header.js';
import { PduHeaderConverter } from '../std_msgs/pdu_cdr_conv_Header.js';
import { Quaternion } from '../geometry_msgs/pdu_jstype_Quaternion.js';
import { PduQuaternionConverter } from '../geometry_msgs/pdu_cdr_conv_Quaternion.js';
import { Time } from '../builtin_interfaces/pdu_jstype_Time.js';
import { PduTimeConverter } from '../builtin_interfaces/pdu_cdr_conv_Time.js';
import { Vector3 } from '../geometry_msgs/pdu_jstype_Vector3.js';
import { PduVector3Converter } from '../geometry_msgs/pdu_cdr_conv_Vector3.js';


export class PduImuConverter {
    /**
     * @param {PduCdrWriter} writer
     * @param { Imu } src
     */
    static to_cdr_body(writer, src) {
        PduHeaderConverter.to_cdr_body(writer, src.header);
        PduQuaternionConverter.to_cdr_body(writer, src.orientation);
        {
            const values = src.orientation_covariance;
            for (let i = 0; i < 9; i++) {
                writer.write_float64(i < values.length ? values[i] : 0.0);
            }
        }
        PduVector3Converter.to_cdr_body(writer, src.angular_velocity);
        {
            const values = src.angular_velocity_covariance;
            for (let i = 0; i < 9; i++) {
                writer.write_float64(i < values.length ? values[i] : 0.0);
            }
        }
        PduVector3Converter.to_cdr_body(writer, src.linear_acceleration);
        {
            const values = src.linear_acceleration_covariance;
            for (let i = 0; i < 9; i++) {
                writer.write_float64(i < values.length ? values[i] : 0.0);
            }
        }
    }

    /**
     * @param {PduCdrReader} reader
     * @param { Imu } dst
     * @returns { Imu }
     */
    static cdr_body_to_js(reader, dst) {
        PduHeaderConverter.cdr_body_to_js(reader, dst.header);
        PduQuaternionConverter.cdr_body_to_js(reader, dst.orientation);
        dst.orientation_covariance = Array.from({ length: 9 }, () => reader.read_float64());
        PduVector3Converter.cdr_body_to_js(reader, dst.angular_velocity);
        dst.angular_velocity_covariance = Array.from({ length: 9 }, () => reader.read_float64());
        PduVector3Converter.cdr_body_to_js(reader, dst.linear_acceleration);
        dst.linear_acceleration_covariance = Array.from({ length: 9 }, () => reader.read_float64());
        return dst;
    }

    /**
     * @param { Imu } src
     * @returns {ArrayBuffer}
     */
    static to_cdr(src) {
        const writer = new PduCdrWriter();
        writer.write_encapsulation();
        this.to_cdr_body(writer, src);
        return writer.get_buf();
    }

    /**
     * @param {ArrayBuffer|ArrayBufferView} cdrPayload
     * @returns { Imu }
     */
    static from_cdr(cdrPayload) {
        const reader = new PduCdrReader(cdrPayload);
        reader.read_encapsulation();
        return this.cdr_body_to_js(reader, new Imu());
    }
}
